File: src/analytics/mlPredictor.ts
import { OHLCVData, MLPrediction } from '../types';
import { technicalIndicators } from './indicators';

export class MLPredictor {
  private features: number[][] = [];
  private targets: number[] = [];
  private isTrained = false;
  private weights: number[] = [];
  private bias = 0;

  predict(ohlcvData: OHLCVData[]): MLPrediction {
    if (!this.isTrained) {
      this.quickTrain(ohlcvData);
    }
    
    if (!this.isTrained || ohlcvData.length < 20) {
      return {
        score: 0.5,
        prediction: 'HOLD',
        confidence: 0.0
      };
    }
    
    const features = this.extractFeatures(ohlcvData);
    if (features.length === 0) {
      return {
        score: 0.5,
        prediction: 'HOLD', 
        confidence: 0.0
      };
    }
    
    const currentFeatures = features[features.length - 1];
    const prediction = this.simpleLinearPredict(currentFeatures);
    
    const bullishProb = Math.max(0, Math.min(1, prediction));
    
    let action: 'BUY' | 'SELL' | 'HOLD' = 'HOLD';
    if (bullishProb > 0.6) action = 'BUY';
    else if (bullishProb < 0.4) action = 'SELL';
    
    return {
      score: bullishProb,
      prediction: action,
      confidence: Math.abs(bullishProb - 0.5) * 2
    };
  }

  private extractFeatures(ohlcvData: OHLCVData[]): number[][] {
    if (ohlcvData.length < 20) return [];
    
    const prices = ohlcvData.map(d => d.close);
    const volumes = ohlcvData.map(d => d.volume);
    const highs = ohlcvData.map(d => d.high);
    const lows = ohlcvData.map(d => d.low);
    
    const rsi = technicalIndicators.calculateRSI(prices);
    const ema5 = technicalIndicators.calculateEMA(prices, 5);
    const ema20 = technicalIndicators.calculateEMA(prices, 20);
    const avgVolume = volumes.reduce((a, b) => a + b, 0) / volumes.length;
    
    const features: number[][] = [];
    
    for (let i = 20; i < ohlcvData.length; i++) {
      const priceChange = (prices[i] - prices[i - 1]) / prices[i - 1];
      const highLowRatio = (highs[i] - lows[i]) / prices[i];
      const volumeRatio = volumes[i] / avgVolume;
      const rsiValue = rsi[i] / 100; // Normalize to 0-1
      const emaRatio = ema5[i] / ema20[i];
      
      if (!isNaN(priceChange) && !isNaN(highLowRatio) && !isNaN(volumeRatio) && 
          !isNaN(rsiValue) && !isNaN(emaRatio)) {
        features.push([
          priceChange,
          highLowRatio,
          Math.min(volumeRatio, 5), // Cap at 5x
          rsiValue,
          emaRatio
        ]);
      }
    }
    
    return features;
  }

  private quickTrain(ohlcvData: OHLCVData[]): void {
    if (ohlcvData.length < 50) return;
    
    const features = this.extractFeatures(ohlcvData);
    const prices = ohlcvData.map(d => d.close);
    
    // Create targets (future price direction)
    const targets: number[] = [];
    for (let i = 20; i < prices.length - 1; i++) {
      const futureReturn = (prices[i + 1] - prices[i]) / prices[i];
      targets.push(futureReturn > 0 ? 1 : 0);
    }
    
    if (features.length !== targets.length || features.length < 10) return;
    
    // Simple linear regression for quick training
    this.weights = new Array(5).fill(0);
    this.bias = 0;
    
    const learningRate = 0.01;
    const epochs = 100;
    
    for (let epoch = 0; epoch < epochs; epoch++) {
      for (let i = 0; i < features.length; i++) {
        const prediction = this.simpleLinearPredict(features[i]);
        const error = targets[i] - prediction;
        
        // Update weights
        for (let j = 0; j < this.weights.length; j++) {
          this.weights[j] += learningRate * error * features[i][j];
        }
        this.bias += learningRate * error;
      }
    }
    
    this.isTrained = true;
  }

  private simpleLinearPredict(features: number[]): number {
    let prediction = this.bias;
    for (let i = 0; i < features.length && i < this.weights.length; i++) {
      prediction += this.weights[i] * features[i];
    }
    
    // Sigmoid activation
    return 1 / (1 + Math.exp(-prediction));
  }
}

export const mlPredictor = new MLPredictor();

File: backend/analytics/ml_predictor.py
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')

class MLPredictor:
    def __init__(self):
        self.model = RandomForestClassifier(n_estimators=50, random_state=42, max_depth=10)
        self.scaler = StandardScaler()
        self.is_trained = False
        self.feature_names = [
            'price_change', 'high_low_ratio', 'volume_ratio',
            'rsi', 'ema_ratio', 'volatility', 'momentum'
        ]
    
    def prepare_features(self, ohlcv_data: pd.DataFrame) -> np.ndarray:
        """Create feature set from OHLCV data"""
        df = ohlcv_data.copy()
        
        # Price features
        df['price_change'] = df['close'].pct_change()
        df['high_low_ratio'] = (df['high'] - df['low']) / df['close']
        df['volume_ratio'] = df['volume'] / df['volume'].rolling(10).mean()
        
        # Technical indicators
        df['rsi'] = self._calculate_simple_rsi(df['close'])
        df['ema_5'] = df['close'].ewm(span=5).mean()
        df['ema_20'] = df['close'].ewm(span=20).mean()
        df['ema_ratio'] = df['ema_5'] / df['ema_20']
        
        # Additional features
        df['volatility'] = df['close'].rolling(10).std() / df['close'].rolling(10).mean()
        df['momentum'] = df['close'] / df['close'].shift(5) - 1
        
        # Select features
        feature_matrix = df[self.feature_names].fillna(0).values
        return feature_matrix
    
    def _calculate_simple_rsi(self, prices: pd.Series, period: int = 14) -> pd.Series:
        """Calculate RSI for feature engineering"""
        delta = prices.diff()
        gain = (delta.where(delta > 0, 0)).rolling(window=period).mean()
        loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()
        rs = gain / loss
        rsi = 100 - (100 / (1 + rs))
        return rsi.fillna(50)
    
    def train_model(self, ohlcv_data: pd.DataFrame) -> bool:
        """Train the ML model on historical data"""
        if len(ohlcv_data) < 100:
            return False
        
        try:
            features = self.prepare_features(ohlcv_data)
            
            # Create target (future price direction)
            future_returns = ohlcv_data['close'].shift(-1) / ohlcv_data['close'] - 1
            target = (future_returns > 0).astype(int)
            
            # Remove last row (no future data) and NaN values
            features = features[:-1]
            target = target[:-1]
            
            # Remove NaN values
            valid_indices = ~np.isnan(features).any(axis=1) & ~np.isnan(target)
            features = features[valid_indices]
            target = target[valid_indices]
            
            if len(features) < 50:
                return False
            
            # Split data for training
            X_train, X_test, y_train, y_test = train_test_split(
                features, target, test_size=0.2, random_state=42
            )
            
            # Scale features
            X_train_scaled = self.scaler.fit_transform(X_train)
            X_test_scaled = self.scaler.transform(X_test)
            
            # Train model
            self.model.fit(X_train_scaled, y_train)
            
            # Validate model
            train_score = self.model.score(X_train_scaled, y_train)
            test_score = self.model.score(X_test_scaled, y_test)
            
            print(f"ML Model trained - Train Score: {train_score:.3f}, Test Score: {test_score:.3f}")
            
            self.is_trained = True
            return True
            
        except Exception as e:
            print(f"Error training ML model: {e}")
            return False
    
    def predict(self, ohlcv_data: pd.DataFrame) -> dict:
        """Make ML prediction (5% weight in final algorithm)"""
        if not self.is_trained:
            # Quick training on available data
            self.train_model(ohlcv_data)
        
        if not self.is_trained or len(ohlcv_data) < 20:
            return {
                'score': 0.5,
                'prediction': 'NEUTRAL',
                'confidence': 0.0
            }
        
        try:
            # Prepare current features
            features = self.prepare_features(ohlcv_data)
            current_features = features[-1:].reshape(1, -1)
            
            # Handle NaN values
            if np.isnan(current_features).any():
                return {
                    'score': 0.5,
                    'prediction': 'NEUTRAL', 
                    'confidence': 0.0
                }
            
            # Scale and predict
            features_scaled = self.scaler.transform(current_features)
            prediction_proba = self.model.predict_proba(features_scaled)[0]
            
            # Get prediction and confidence
            if len(prediction_proba) > 1:
                bullish_prob = prediction_proba[1]
            else:
                bullish_prob = 0.5
            
            # Determine action
            if bullish_prob > 0.6:
                prediction = 'BUY'
            elif bullish_prob < 0.4:
                prediction = 'SELL'
            else:
                prediction = 'HOLD'
            
            confidence = abs(bullish_prob - 0.5) * 2
            
            return {
                'score': bullish_prob,
                'prediction': prediction,
                'confidence': confidence
            }
            
        except Exception as e:
            print(f"Error making ML prediction: {e}")
            return {
                'score': 0.5,
                'prediction': 'NEUTRAL',
                'confidence': 0.0
            }

# Global ML predictor instance
ml_predictor = MLPredictor()